pkg/controller/datavolume-controller.go

```diff
@@ -373,194 +383,319 @@ func (r *DatavolumeReconciler) Reconcile(_ context.Context, req reconcile.Reques
-				}
+			if !done {
+				return reconcile.Result{},
+					r.updateSmartCloneStatusPhase(cdiv1.ExpansionInProgress, datavolume, nil)
 			}
-		}
-	}
 
-	if doSmartClone {
-		if isCrossNamespaceClone(datavolume) && datavolume.Status.Phase == cdiv1.Succeeded {
-			if err := r.cleanupTransfer(log, datavolume, transferName); err != nil {
+			// trigger transfer and next reconcile should have pvcExists == true
+			tmpPVC.Annotations[annReadyForTransfer] = "true"
+			tmpPVC.Annotations[AnnPopulatedFor] = datavolume.Name
+			if err := r.client.Update(context.TODO(), tmpPVC); err != nil {
 				return reconcile.Result{}, err
 			}
 
-			// done, done
+			return reconcile.Result{},
+				r.updateSmartCloneStatusPhase(cdiv1.NamespaceTransferInProgress, datavolume, nil)
+		} else {
 			return reconcile.Result{}, nil
 		}
+	}
 
-		if pvc.Annotations[AnnCloneOf] != "true" {
-			return reconcile.Result{}, nil
+	if datavolume.Status.Phase == cdiv1.NamespaceTransferInProgress {
+		return reconcile.Result{}, nil
+	}
+
+	r.log.V(3).Info("Smart-Clone via Snapshot is available with Volume Snapshot Class",
+		"snapshotClassName", snapshotClassName)
+
+	newSnapshot := newSnapshot(datavolume, pvcName, snapshotClassName)
+	if err := setAnnOwnedByDataVolume(newSnapshot, datavolume); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	nn := client.ObjectKeyFromObject(newSnapshot)
+	if err := r.client.Get(context.TODO(), nn, newSnapshot.DeepCopy()); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, err
 		}
 
-		// expand for non-namespace case
-		done, err := r.expand(log, datavolume, pvc, pvcSpec)
+		inUse, err := r.sourceInUse(datavolume)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
-		if !done {
-			return reconcile.Result{},
-				r.updateSmartCloneStatusPhase(cdiv1.ExpansionInProgress, datavolume, pvc)
+		populated, err := r.isSourcePVCPopulated(datavolume)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if inUse || !populated {
+			return reconcile.Result{Requeue: true},
+				r.updateSmartCloneStatusPhase(cdiv1.CloneScheduled, datavolume, nil)
 		}
 
-		// done
-		return reconcile.Result{}, r.updateSmartCloneStatusPhase(cdiv1.Succeeded, datavolume, pvc)
+		if err := r.client.Create(context.TODO(), newSnapshot); err != nil {
+			if k8serrors.IsAlreadyExists(err) {
+				return reconcile.Result{}, nil
+			}
+
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{},
+			r.updateSmartCloneStatusPhase(cdiv1.SnapshotForSmartCloneInProgress, datavolume, nil)
 	}
 
-	// Finally, we update the status block of the DataVolume resource to reflect the
-	// current state of the world
-	return r.reconcileDataVolumeStatus(datavolume, pvc)
+	return reconcile.Result{}, nil
+}
+
+func (r *DatavolumeReconciler) reconcileSmartCloneForExistingPvc(log logr.Logger, datavolume *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim, pvcSpec *corev1.PersistentVolumeClaimSpec, transferName string) (reconcile.Result, error) {
```

Comment:
Isn't this whole function just post clone processing (transfer and/or resize) for either CSI clone or smart clone. I think the function name should reflect its purpose. It is not restricted to smart clone. And some of the updateStatus functions should probably also be renamed to reflect this.

Reply:
As of this PR the updateStatus you mentioned is only used by smartClone, but I'll for sure rename and reuse in the CSI-clone PR. 

The same is true for `reconcileSmartCloneForExistingPvc`, and I agree it needs a better name.

Reply:
Okay, I will hold you to that in the other PR then.